File: Data Structures & Algorithms/balanced-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        self.balanced = True

        def dfs(root):
            if root is None:
                return 0
            leftval = dfs(root.left) 
            rightval = dfs(root.right)

            if abs(leftval-rightval) > 1:
                self.balanced = False
            return max(leftval,rightval) + 1

        dfs(root)
        return self.balanced


        

File: Data Structures & Algorithms/binary-tree-diameter/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <algorithm>
class Solution {
private:
int mx_diameter = 0;
    int height(TreeNode* node) {
        if (node == nullptr) {
            return 0;
        }

        int left_height = height(node->left);
        int right_height = height(node->right);

        mx_diameter = std::max(mx_diameter, left_height + right_height);

        return 1 + std::max(left_height, right_height);
    }
public:
    int diameterOfBinaryTree(TreeNode* root) {
        height(root);
        return mx_diameter;
    }
};
